instrumentation/spring-rabbit/src/main/java/brave/spring/rabbit/TracingRabbitListenerAdvice.java

```diff
@@ -72,40 +76,53 @@
    * Message)}
    */
   @Override public Object invoke(MethodInvocation methodInvocation) throws Throwable {
-    Message message = (Message) methodInvocation.getArguments()[1];
-    MessageConsumerRequest request = new MessageConsumerRequest(message);
+    List<Message> messages = new ArrayList<>();
+    if (methodInvocation.getArguments()[1] instanceof List) {
+      messages.addAll((Collection<? extends Message>) methodInvocation.getArguments()[1]);
+    } else {
+      messages.add((Message) methodInvocation.getArguments()[1]);
+    }
+
+    List<Span> listenerSpans = new ArrayList<>(messages.size());
+    for (Message message : messages) {
+      MessageConsumerRequest request = new MessageConsumerRequest(message);
 
-    TraceContextOrSamplingFlags extracted =
-      springRabbitTracing.extractAndClearTraceIdHeaders(extractor, request, message);
+      TraceContextOrSamplingFlags extracted =
+        springRabbitTracing.extractAndClearTraceIdHeaders(extractor, request, message);
 
-    // named for BlockingQueueConsumer.nextMessage, which we can't currently see
-    Span consumerSpan = springRabbitTracing.nextMessagingSpan(sampler, request, extracted);
-    Span listenerSpan = tracer.newChild(consumerSpan.context());
+      // named for BlockingQueueConsumer.nextMessage, which we can't currently see
+      Span consumerSpan = springRabbitTracing.nextMessagingSpan(sampler, request, extracted);
```

Comment:
This is an interesting situation as we want one consumer span for the batch I think.. also, there's only one listener, so should only be one listener span!

The tricky bit is that we have an ambiguous situation. For example, we may have different trace contexts in the batch of messages. Consider 5 messages, and two of them have "b3" header in them.

This "join" type operation forces us to choose which of these possibly multiple traces should be the trace of the listener.

Another option is like you've done, which is to start a span for every message, but since only one can be the "current trace" we still I think should pick one. We don't yet have a link model, but meanwhile yeah I feel like we should have only 1 span started for the listener, because only one child started in a listener could be attached to a trace.

cc @jorgheymans @anuraaga @jeqo for feedback on this.

Reply:
I'd generally go with span per message since batching is an implementation detail of how the messages are handled. Many user requests to buy a sticker result in wanting to write a money transaction to a database, queues are great for supporting retries idempotently to make that reliable but each transaction is still independent from each other. To me, each message is very much like the request to an API server, so the messages without any header in them would just be new traces.

Note, there might be other cases where this isn't really the cases (the messages are all related and part of some larger type of processing, maybe analytics?) in which case single span could make sense too.

Reply:
I realized, actually this change is probably related to my second paragraph - if it's a configuration change to switch from one message per listener to one batch per listener, that means a user is sort of opting into that behavior, presumably since there's a semantic relationship among the batch. So just one span does seem to make sense, but picking the trace is problematic.

Reply:
So the solution is keep only one consumer span and one listener span  (maybe from the first message)？

Reply:
so far, I don't see a better way. for example, if we see any messages we can start and finish a consumer span, but we keep a reference to only the first consumer span, if exists use that as the listener's parent

Reply:
@adriancole Please re-review. Thanks.